assets/src/stories-editor/components/editor-carousel/indicator.js

```diff
@@ -29,27 +57,109 @@ const Indicator = ( { pages, currentPage, onClick } ) => {
 	/* translators: %s: Page number */
 	const toolTip = ( pageNumber ) => sprintf( __( 'Go to page %s', 'amp' ), pageNumber );
 
+	const [ isDragging, setIsDragging ] = useState( false );
+	const { movePageToPosition, initializePageOrder } = useDispatch( 'amp/story' );
+	const { getBlockOrder, getBlockIndex } = useSelect( ( select ) => {
+		return select( 'core/block-editor' );
+	} );
+
 	return (
 		<ul className="amp-story-editor-carousel-item-list">
 			{ pages.map( ( page, index ) => {
 				const className = page.clientId === currentPage ? 'amp-story-editor-carousel-item amp-story-editor-carousel-item--active' : 'amp-story-editor-carousel-item';
+				const blockElementId = `amp-story-editor-carousel-item-${ page.clientId }`;
+				const transferData = {
+					type: 'indicator',
+					srcIndex: getBlockIndex( page.clientId ),
+					srcClientId: page.clientId,
+				};
+
+				const getInsertIndex = ( position ) => {
+					const dstIndex = getBlockIndex( page.clientId );
+					if ( page.clientId !== undefined ) {
+						return position.x === 'left' ? dstIndex : dstIndex + 1;
+					}
+
+					return undefined;
+				};
+
+				const onDrop = ( event, position ) => {
+					const { srcClientId, srcIndex, type } = parseDropEvent( event );
+
+					const isIndicatorDropType = ( dropType ) => dropType === 'indicator';
+					const isSameBlock = ( src, dst ) => src === dst;
+
+					if ( ! isIndicatorDropType( type ) || isSameBlock( srcClientId, page.clientId ) ) {
+						return;
+					}
+
+					const positionIndex = getInsertIndex( position );
+					const insertIndex = srcIndex < index ? positionIndex - 1 : positionIndex;
+					movePageToPosition( srcClientId, insertIndex );
+				};
 
 				return (
-					<li key={ page.clientId } className={ className }>
-						<Tooltip text={ toolTip( index + 1 ) }>
-							<Button
-								onClick={ ( e ) => {
-									e.preventDefault();
-									onClick( page.clientId );
-								} }
-								disabled={ page.clientId === currentPage }
-							>
-								<span className="screen-reader-text">
-									{ label( index + 1 ) }
-								</span>
-							</Button>
-						</Tooltip>
-					</li>
+					<Draggable
+						key={ page.clientId }
+						elementId={ blockElementId }
+						transferData={ transferData }
+						onDragStart={ () => {
+							setIsDragging( page.clientId );
+							initializePageOrder( getBlockOrder() );
+						} }
+						onDragEnd={ () => {
+							setIsDragging( false );
+						} }
+					>
+						{
+							( { onDraggableStart, onDraggableEnd } ) => (
+								<>
+									<li
+										className={ className }
+									>
+										<div
+											onDragStart={ onDraggableStart }
+											onDragEnd={ onDraggableEnd }
+											draggable
+											className="amp-story-editor-carousel-item-wrapper"
+											id={ `amp-story-editor-carousel-item-${ page.clientId }` }
+										>
+											{ page.clientId !== currentPage && (
```

Comment:
Render the button to a variable above this return:

```
const button = (
	<Button ... />
);
```

And then change this to a ternary:

```
{ isCurrentPage ? <Tooltip ...>{ button }</Tooltip> : button }
```